server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+				sprintf( __( 'The %s argument is required and must be a string.', 'performance-lab' ), '$args["access_cap"]' ),
+				''
+			);
+			return;
+		}
+
+		$this->registered_metrics[ $metric_slug ]      = new Perflab_Server_Timing_Metric( $metric_slug );
+		$this->registered_metrics_data[ $metric_slug ] = $args;
+
+		// If the current user has already been determined and they lack the necessary access,
+		// do not even attempt to calculate the metric.
+		if ( did_action( 'set_current_user' ) && ! current_user_can( $args['access_cap'] ) ) {
+			return;
+		}
+
+		// Otherwise, call the measuring callback and pass the metric instance to it.
+		call_user_func( $args['measure_callback'], $this->registered_metrics[ $metric_slug ] );
+	}
+
+	/**
+	 * Checks whether the given metric has been registered.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @return bool True if registered, false otherwise.
+	 */
+	public function has_registered_metric( $metric_slug ) {
+		return isset( $this->registered_metrics[ $metric_slug ] ) && isset( $this->registered_metrics_data[ $metric_slug ] );
+	}
+
+	/**
+	 * Outputs the Server-Timing header.
+	 *
+	 * This method must be called before rendering the page.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function send_header() {
+		if ( headers_sent() ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The method must be called before headers have been sent.', 'performance-lab' ),
+				''
+			);
+			return;
+		}
+
+		/**
+		 * Fires right before the Server-Timing header is sent.
+		 *
+		 * This action is the last possible point to register a Server-Timing metric.
+		 *
+		 * @since n.e.x.t
+		 */
+		do_action( 'perflab_server_timing_send_header' );
+
+		$header_value = $this->get_header_value();
+		if ( ! $header_value ) {
+			return;
+		}
+
+		header( sprintf( 'Server-Timing: %s', $header_value ), false );
+	}
+
+	/**
+	 * Gets the value for the Server-Timing header.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string The Server-Timing header value.
+	 */
+	public function get_header_value() {
+		// Get all metric header values, as long as the current user has access to the metric.
+		$metric_header_values = array_filter(
+			array_map(
+				function( Perflab_Server_Timing_Metric $metric ) {
+					$value = $metric->get_value();
+					if ( null === $value ) {
+						return null;
+					}
+
+					// Check the registered capability here to ensure no metric without access is exposed.
+					if ( ! current_user_can( $this->registered_metrics_data[ $metric->get_slug() ]['access_cap'] ) ) {
+						return null;
+					}
+
+					return $this->format_metric_header_value( $metric );
+				},
+				$this->registered_metrics
+			),
+			function( $value ) {
+				return null !== $value;
+			}
+		);
+
+		return implode( ', ', $metric_header_values );
+	}
+
+	/**
+	 * Returns whether an output buffer should be used to gather Server-Timing metrics during template rendering.
```

Comment:
Might be useful to add one line doc here indicating **why** you might want to do this